crates/stream-download/src/storage/mod.rs

```diff
@@ -9,6 +9,142 @@ pub mod memory;
 #[cfg(feature = "temp-storage")]
 pub mod temp;
 
+/// Represents a content length that can change during processing.
+///
+/// This structure holds:
+/// - `reported`: The length of the content as reported by the server (e.g., Content-Range header).
+/// - `gathered`: The actual length of the content after processing (e.g., decryption,
+///   decompression), or `None` if the final value is not yet known (still being calculated or
+///   accumulated).
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub struct DynamicLength {
+    /// The length of the content as reported by the server, for example, via Content-Range header.
+    pub reported: u64,
+    /// The actual content length gathered after all processing.
+    /// This remains `None` until the entire content is processed and its final length determined.
+    pub gathered: Option<u64>,
+}
+
+/// Describes the possible states for the length of a stream or content resource.
+///
+/// `ContentLength` distinguishes between content with a fixed, knowable size,
+/// content whose final size is determined only after processing,
+/// and content where the length is unknown.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum ContentLength {
```

Comment:
This feels like it probably belongs in the `source` module rather than `storage`.